byron/ledger/executable-spec/src/Ledger/UTxO/Generators.hs

```diff
@@ -0,0 +1,273 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Ledger.UTxO.Generators where
+
+import Control.Applicative (empty)
+import Data.Bitraversable (bitraverse)
+import qualified Data.Map.Strict as M
+
+import Ledger.Core hiding (Range, range)
+import Ledger.UTxO hiding (utxo)
+
+import Hedgehog (Gen, Property, Range, (===), assert, forAll, property)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+import Hedgehog.Internal.Gen
+  (atLeast, ensure, mapGenT, runDiscardEffect, toTree, toTreeMaybeT)
+import Hedgehog.Internal.Tree (NodeT(..), TreeT(..), treeValue)
+import qualified Hedgehog.Internal.Tree as Tree
+
+
+--------------------------------------------------------------------------------
+-- Initial TxOut generator
+--------------------------------------------------------------------------------
+
+-- | Generate a set of initial @TxOut@s from a set of @Addr@s
+genInitialTxOuts :: [Addr] -> Gen [TxOut]
+genInitialTxOuts = Gen.filter (not . null)
+  . genTraverseSubsequence (\a -> TxOut a <$> genLovelace)
+
+genLovelace :: Gen Lovelace
+genLovelace = Lovelace . fromIntegral <$> Gen.word32 (Range.linear 1 10000)
+
+-- | Generate a subsequence of a list of values and traverse the subsequence
+--   with a generator producer
+genTraverseSubsequence :: (a -> Gen b) -> [a] -> Gen [b]
+genTraverseSubsequence genA as =
+  mapGenT (TreeT . (interleaveTreeT . nodeValue =<<) . runTreeT) $ do
+    sub <- Gen.subsequence as
+    traverse (toTreeMaybeT . genA) sub
+
+-- | Generate a list using @genTraverseSubsequence@
+genList :: Range Int -> Gen a -> Gen [a]
+genList range gen = Gen.sized $ \size ->
+  ensure (atLeast $ Range.lowerBound size range) $ genTraverseSubsequence
+    (const gen)
+    (replicate (Range.upperBound size range) ())
+
+-- | Temporarily defined here until hedgehog exposes this function
+interleaveTreeT :: Monad m => [TreeT m a] -> m (NodeT m [a])
+interleaveTreeT = fmap Tree.interleave . traverse runTreeT
+
+
+--------------------------------------------------------------------------------
+-- Tx generator
+--------------------------------------------------------------------------------
+
+-- | Generate a valid transaction from a given @UTxO@
+genTxFromUTxO :: Ord id => id -> [Addr] -> UTxO id -> Gen (Tx id)
+genTxFromUTxO placeholderId addrs utxo =
+  uncurry (Tx placeholderId) <$> Gen.filter
+    (not . null . fst)
+    (genInputOutput
+      (M.keys $ unUTxO utxo)
+      (maybe 0 (unLovelace . value) . flip M.lookup (unUTxO utxo))
+      (fmap (. Lovelace) $ TxOut <$> Gen.element addrs)
+      (unLovelace . value)
+      (\f out -> out { value = Lovelace . f . unLovelace $ value out })
+    )
+
+
+-- | A property to test that the entire shrink tree generated by
+--   @genInputOutput@ maintains the invariant that the inputs and outputs have
+--   equal sums
+propGenInputOutput :: Property
+propGenInputOutput = property $ do
+  ins <- forAll $ Gen.list (Range.linear 0 5) (Gen.integral (Range.linear 0 5))
```

Comment:
And like always I'm curious about why did we choose an upper bound of `5` instead of say `100`.

Reply:
I second that (although it is probably ok here)
how about using and exponential distribution for small values and an equal distribution "in between" and another exponential distribution for large values to some maximal value? I think that might generally be a useful generator for Lovelace values

Reply:
The reason is that large values generate *very large* shrink trees! This test forces the full tree, so no laziness to keep things sane, even with 10 there things get a bit crazy.

Reply:
Oh, that makes sense. Does it make sense to document what you said above so that people like me don't make stupid mistakes? :grin: